_posts/2023-03-26-Kafka-IaC.md

```diff
@@ -0,0 +1,249 @@
+List of Topics:
+<topics that are created>
+List of ACLs:
+<acls that are created>
+List of Principles:
+List of Connectors:
+List of KSQL Artifacts:
+Kafka Topology updated
+```
+
+Want a quick start? checkout our sample JulieOps repo in [here].
+
+[julie-ops]: https://julieops.readthedocs.io/en/latest/#
+[Topics]: https://julieops.readthedocs.io/en/latest/futures/what-topic-management.html
+[Handling schemas]: https://julieops.readthedocs.io/en/latest/futures/what-schema-management.html
+[Access Control]: https://julieops.readthedocs.io/en/latest/futures/what-acl-management.html
+[ksql artifacts]: https://julieops.readthedocs.io/en/latest/futures/what-ksql-management.html
+[topologies]: https://julieops.readthedocs.io/en/latest/the-descriptor-files.html?highlight=topology
+[here]: https://github.com/Platformatory/kafka-cd-julie
+
+## Pulumi
+
+Choosing the right IaC(Infrastructure as a Code) tool is important. Each tool has its own pros and cons. We have already seen a brief about IaC in the previous section. Here, we will be provisioning the Confluent Cloud Topics and Connectors with Pulumi. Though Pulumi supports a wide variety of programming languages like Python, Typescript, Go, C3, Java & YAML we will be using Typescript in this blog.
+
+### Provisioning Kafka Topics
+
+There are a couple of mandatory configs that are needed to create Kafka Topics. To begin with we need cluster arguments on which the infrastructure needs to be provisioned.
+
+```
+let clusterArgs: KafkaTopicKafkaCluster = {
+  id: cluster_id,
+};
+```
+
+Then, we will be needing the Kafka credentials the API Key & API Secret.
+
+```
+let clusterCredentials: KafkaTopicCredentials = {
+  key: kafka_api_key,
+  secret: kafka_api_secret,
+};
+```
+Then comes the topic configs. There are various Kafka topic configurations. Read about them [here](https://kafka.apache.org/documentation/#topicconfigs)
+
+```
+  let topic_args: KafkaTopicArgs = {
+    kafkaCluster: clusterArgs,
+    topicName: topicName.toLowerCase(),
+    restEndpoint: rest_endpoint,
+    credentials: clusterCredentials,
+    config: {
+      ["retention.ms"]: "-1",
+      ["retention.bytes"]: "-1",
+      ["num.partitions"]: "6",
+    },
+  };
+```
+
+Finally the creation of topics
+```
+const topics = new confluent.KafkaTopic(
+    topicNames[i].toLowerCase(),
+    topic_args
+  );
+```
+
+Save the above file to an `index.ts` and set the confluent cloud cluster credentials using `pulumi config set confluentcloud:cloudApiKey <cloud api key> --secret && pulumi config set confluentcloud:cloudApiSecret <cloud api secret> --secret`. It is important to pass the `--secret` flag to the config else the secrets will not be masked on the Pulumi infrastructure config files. On setting the credentials, pulumi will prompt for a stack to be selected. Select the stack if it already exists, else create a new stack.
+
+Once the credentials are set run `pulumi up` command to provision the topics in confluent cloud.
+
```

Comment:
We can highlight the difference among the different IaC tools with respect to configurations, defining resources, storing state etc.